From 5b58d823c5d01c535a9b1f36636fc45a1bed46a8 Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 5 Sep 2014 16:36:26 -0400
Subject: [infiniband] ulp/isert: Wait for proper cleanup before unloading

Message-id: <1409934989-29941-29-git-send-email-agrover@redhat.com>
Patchwork-id: 91482
O-Subject: [RHEL7.1 PATCH e-stor 28/31] Target/iser: Wait for proper cleanup before unloading
Bugzilla: 1129387
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

From: Sagi Grimberg <sagig@mellanox.com>

commit f5ebec9629cf78eeeea4b8258882a9f439ab2404 upstream.

disconnected_handler works are scheduled on system_wq.
When attempting to unload, first make sure all works
have cleaned up.

Signed-off-by: Sagi Grimberg <sagig@mellanox.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/isert/ib_isert.c b/drivers/infiniband/ulp/isert/ib_isert.c
index 09f94b2..0aa33cd 100644
--- a/drivers/infiniband/ulp/isert/ib_isert.c
+++ b/drivers/infiniband/ulp/isert/ib_isert.c
@@ -2334,6 +2334,7 @@ destroy_rx_wq:
 
 static void __exit isert_exit(void)
 {
+ flush_scheduled_work();
  kmem_cache_destroy(isert_cmd_cache);
  destroy_workqueue(isert_comp_wq);
  destroy_workqueue(isert_rx_wq);
-- 
1.7.1